### DISCUSSION TITLE
Proxy Grafana Live WebSocket connections

### DISCUSSION THREAD

**Discussion by sinedoOo** [nginx/unit#1120] [Support]
In official guide of Grafana there is example configuration for Nginx.

```
# this is required to proxy Grafana Live WebSocket connections.
map $http_upgrade $connection_upgrade {
  default upgrade;
  '' close;
}

upstream grafana {
  server localhost:3000;
}

server {
  listen 80;
  root /usr/share/nginx/html;
  index index.html index.htm;

  location / {
    proxy_set_header Host $http_host;
    proxy_pass http://grafana;
  }

# Proxy Grafana Live WebSocket connections.
  location /api/live/ {
    proxy_http_version 1.1;
    proxy_set_header Upgrade $http_upgrade;
    proxy_set_header Connection $connection_upgrade;
    proxy_set_header Host $http_host;
    proxy_pass http://grafana;
  }
}
```

Unfortunately I cannot force the live WebSockets to work. Any help how I can configure it in Unit? Simple config with proxy works except broken WebSockets requests.

**Comment by tippexs** [comment]
Hi @sinedoOo thanks for reaching out but this is the issue tracker for NGINX Unit. Please repost your question either in our NGINX Community Slack or on the NIGNX mailinglist. For more information see the [NGINX Support Page](https://nginx.org/en/support.html)

> Unfortunately I cannot force the live WebSockets to work. 
Please make sure you are providing error messages / logs as well.

**Reply by sinedoOo** [reply]
I’m super confused now. My issue is connected to Unit, not Nginx. I asked how I can configure Unit to proxy WebSocket requests, because I didn’t find this information in official documentation. And posted example config from Nginx, because I found it in official Grafana guide.

**Reply by tippexs** [reply]
Sorry for the confusion @sinedoOo 
The Proxy-Module in NGINX Unit is just a very simple implementation of proxying to other HTTP endpoints. Currently there is no support of adding headers that will be send towards the upstream. I have created a simple NodeJS based Proxy that will be able to do this if this helps but long story short, for Proxy use cases we recommend to use NGINX. Sorry again for the confusion.

I have re-opened the discussion so we can find a proper solution for your issue.

**Reply by sinedoOo** [reply]
Ok, so basically there is no way to support WebSockets pass through Unit, right?

**Reply by tippexs** [reply]
No as the proxy functionality in Unit does not support settings headers that will be shard with the upstream. In this case I would recommend to build the proxy stack using Node,Rust or Python. I know this is not ideal for know but would be best solution. While thinking about it, having a proxy application could be quite beneficial. 

In the meantime Nginx would be the better option for this use case.
